docs/internal/compiler-steps.md

```diff
@@ -0,0 +1,119 @@
+# Compiler Steps
+
+## Overview
+
+This is a preliminary study on each compiler step to produce executable objects from Verona code.
+The initial objective of the Verona compiler is to generate a single executable from a number of files, directories and existing foreing objects.
+Loadable code and separate compilation processes are not taken into account in this document but are under discussion and may be introduced in the future.
+
+### Verona Sources
+
+The Verona compiler treats all files inside the same directory as a single module.
+Point the compiler to the directory and it will compile all (Verona) files together and include other modules (via `use` or `type`) in the same way.
+
+The end result is a single meta-module containing all modules: the main directory and all included modules from any module.
+This final module must contain a `Main` class with a `main(env : Env) : I32` method on any of the sub-modules, that will be the entry point for the program.
+Having zero or more than one `Main` classes is a compiler error.
+
+### Foreign Code
+
+Verona can also interoperate with foreign code, via sandbox regions and remote calls into the foreign object (archive, shared object, etc).
+
+The main goal of the interoperability layer is to use _functionality in foreign modules_, not to use _foreign code_ inside Verona.
+
+Check the `interop.md` document for more details.
+
+Interoperability code can be found at `src/interop`.
+
+## Parser
+
+The parser is responsible for:
+* Understanding how to navigate directory structures to include all reachable code
+* Recognising foreign calls and trigger parsing of foreing code
+* Check syntax and semantics correctness of all sources
+* Parse the code as one meta-module and emit A-normal form AST for the next stage
+
+The final product of the parser is a fully checked concrete AST that the MLIR layer can consume as is (with potential additional helper structures) and generate plain MLIR without additional checks.
+
+The stages (and status) of the parser are:
+* (mostly done) Lex/Parse (+ syntax check)
+* (mostly done) Resolve symbols
+* (in progress) Type inference + Generics + Reification
+* (in progress) Reachability analysis & code elision
+* (mostly done) Emit A-normal form + Symbol table
+
+The parser code can be found at `src/parser`.
+
+## MLIR
+
+The MLIR stage takes in the AST and symbol table above and lowers to the following MLIR dialects:
+* Standard: Overall structure (functions, modules, etc) + existing operations
+* LLVM: Structures, memory management, remaining operations
+* Region (new): A dialect that tags objects into regions + creation & destruction semantics.
+
+With the (high-level) IR above, the following MLIR passes are executed:
+* Region and alias analysis
+* Heap to stack transformation
+* Foreign code RPC generation
```

Comment:
There are a few other things that we'd like to do here (not necessarily in MVP):

* Early inlining (so that we collapse user-defined control flow and elide lambdas early).  Inlining lambdas ideally happens while lambdas and their captures are still abstract and we don't have to go and re-infer things and then elide any lambda-specific memory-management operations.
* Post-inlining type inference, so we can refine the types after inlining.
* `match`-elision so that if we have more refined types we can remove match expressions that we now know are unreachable.

A couple more on the nice-to-have path:

- Union type representation selection (the later we do this, the smaller the set of types in the unions)
- Selector colouring (if we defer this until after we've done post-inlining type inference then we should end up with a smaller set to inlin.

Reply:
> Early inlining (so that we collapse user-defined control flow and elide lambdas early).

I think part of this will be done by the parser (by reducing control flow to a small subset) and the lowering (by converting control flow into basic block flow.

> Post-inlining type inference

Currently, type inference is *completely* done at the parser. Translating the mechanism (or abstracting the API to be used in MLIR) may not be trivial. If the types are all concrete, then whatever is left can be handled by LLVM's own inlining later.

> Union type representation selection
> Selector colouring

Yes! I forgot about those ones. I'll add to the document. Though, selector colouring is currently being done in the parser, too.

Reply:
Added the missing ones in the right places.